maraboupy/MarabouNetworkNNet.py

```diff
@@ -263,31 +563,346 @@ def buildEquations(self):
+            outputs = np.maximum(np.dot(weights[-1], inputsNorm) + biases[-1].reshape((len(biases[-1]), 1)), 0)
+        else:
+            outputs = np.dot(weights[-1], inputsNorm) + biases[-1].reshape((len(biases[-1]), 1))
+
+        # Undo output normalization
+        if (normalize_outputs):
+            for i in range(outputSize):
+                for j in range(numInputs):
+                    outputs[i, j] = outputs[i, j] * self.outputRange + self.outputMean
+
+        return outputs.T
+
+    def evaluateNetworkToLayer(self, inputs, last_layer=-1, normalize_inputs=False, normalize_outputs=False,
+                               activate_output_layer=False):
+
+        """
+        Evaluate the network directly, without Marabou
+        Computes the output at a given layer (outptut layer by default)
+
+
+        Args:
+            inputs (list of floats): Network inputs to be evaluated
+
+        Returns:
+            (numpy array of floats): Network output
+
+       """
+
+        numLayers = self.numLayers
+        inputSize = self.inputSize
+        outputSize = self.outputSize
+        biases = self.biases
+        weights = self.weights
+        mins = self.inputMinimums
+        maxes = self.inputMaximums
+        means = self.inputMeans
+        ranges = self.inputRanges
+
+        # The default output layer is the last (output) layer
+        if last_layer == -1:
+            last_layer = numLayers
+
+        # Prepare the inputs to the neural network
+        if (normalize_inputs):
+            inputsNorm = np.zeros(inputSize)
+            for i in range(inputSize):
+                if inputs[i] < mins[i]:
+                    inputsNorm[i] = (mins[i] - means[i]) / ranges[i]
+                elif inputs[i] > maxes[i]:
+                    inputsNorm[i] = (maxes[i] - means[i]) / ranges[i]
+                else:
+                    inputsNorm[i] = (inputs[i] - means[i]) / ranges[i]
+        else:
+            inputsNorm = inputs
+
+        # Evaluate the neural network
+        for layer in range(last_layer - 1):
+            inputsNorm = np.maximum(np.dot(weights[layer], inputsNorm) + biases[layer], 0)
+
+        layer = last_layer - 1
+
+        if (activate_output_layer):
+            outputs = np.maximum(np.dot(weights[layer], inputsNorm) + biases[layer], 0)
+        else:
+            outputs = np.dot(weights[layer], inputsNorm) + biases[layer]
+
+        # Undo output normalization
+        if (normalize_outputs):
+            if last_layer == numLayers:
+                output_mean = self.outputMean
+                output_range = self.outputRange
+            else:
+                output_mean = means[layer - 1]
+                output_range = ranges[layer - 1]
+
+            for i in range(outputSize):
+                outputs[i] = outputs[i] * output_range + output_mean
+
+        return outputs
+
+    def evaluateNetworkFromLayer(self, inputs, first_layer=0, normalize_inputs=False, normalize_outputs=False,
```

Comment:
What is the use case for this function? Marabou equations are currently not constructed to start or end in an intermediate layer, though that functionality could be added by adding arguments "first_layer" and "last_layer" to the __init__ function. Without that functionality implemented, I don't see how this function is useful for Marabou.

Reply:
I use this just for evaluating the network between given layers, not evaluating Marabou equations. As I have mentioned in a previous comment, this is very helpful for invariant e.g. search. 

Thanks for all the comments, very helpful. I think I've addressed all your concerns, would you mind having another look?

Reply:
No problem, thanks for the work done to add new features! I'll take another look through now.

Reply:
I have now resolved some conflicts that resulted from #317 , and also adjusted docustrings for the new methods to match the new style. 

In addition, I have noticed that there are two ways of creating a Marabou input query from the python interface: one by calling the getInputQuery() method of MarabouNetwork class, and the other one by asking marabou to create one directly (via the interface with the marabou InputQuery class in MarabouCore). This worried me a little, so I added a test to test_query.py that compares these two input queries (by saving them both to temporary files and comparing these files). 

@kjulian3, can you please have another look? I prefer to merge this PR soon, before new conflicts arise. :) 

Thanks!